Index: src/test/groovy/bugs/Groovy7037Bug.groovy
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/test/groovy/bugs/Groovy7037Bug.groovy	(revision )
+++ src/test/groovy/bugs/Groovy7037Bug.groovy	(revision )
@@ -0,0 +1,40 @@
+/*
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package groovy.bugs
+
+class Groovy7037Bug extends GroovyTestCase {
+    void testGetAtShouldWorkWithVarArgs() {
+        assertScript '''
+            class Test {
+                def getAt(Integer x, Long y) { "A: $x $y" }
+                def getAt(Integer x, Long... y) { "B: $x ${y.join(' ')}" }
+            }
+
+            def t = new Test()
+            assert t[2, 4L] == 'A: 2 4'
+            def items = [2, 4L]
+            assert t[*items] == 'A: 2 4'
+            assert t.getAt(2, 4L) == 'A: 2 4'
+            assert t[2] == 'B: 2 '
+            assert t.getAt(2) == 'B: 2 '
+            assert t.getAt(2, 4L, 5L) == 'B: 2 4 5'
+            assert t[2, 4L, 5L] == 'B: 2 4 5'
+        '''
+    }
+}
Index: src/main/org/codehaus/groovy/classgen/asm/BinaryExpressionHelper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/main/org/codehaus/groovy/classgen/asm/BinaryExpressionHelper.java	(date 1472272655000)
+++ src/main/org/codehaus/groovy/classgen/asm/BinaryExpressionHelper.java	(revision )
@@ -242,6 +242,15 @@
             if (controller.getCompileStack().isLHS()) {
                 evaluateEqual(expression, false);
             } else {
+                Expression arguments = expression.getRightExpression();
+                if (arguments instanceof ListExpression) {
+                    Expression newArgs = new ArgumentListExpression(((ListExpression) arguments).getExpressions());
+                    newArgs.setSourcePosition(arguments);
+                    controller.getInvocationWriter().makeCall(
+                            expression, expression.getLeftExpression(), new ConstantExpression("getAt"),
+                            newArgs, InvocationWriter.invokeMethod, false, false, false);
+                    break;
+                }
                 evaluateBinaryExpression("getAt", expression);
             }
             break;
