Index: build.gradle
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- build.gradle	(revision 8301ba03d2a38b36b9f718db54f46b935e1aba5d)
+++ build.gradle	(revision )
@@ -94,6 +94,7 @@
         }
         jcenter()
         maven { url 'http://dl.bintray.com/melix/thirdparty-apache' } // openbeans
+        maven { url 'https://repository.apache.org/snapshots/' }
     }
 
     apply plugin: 'groovy'
@@ -168,7 +169,7 @@
     eclipseOsgiVersion = '3.9.1-v20140110-1610'
     gparsVersion = '1.2.1'
     gradleVersion = '3.5.1'
-    ivyVersion = '2.4.0'
+    ivyVersion = '2.5.0-SNAPSHOT'
     jansiVersion = '1.13'
     jarjarVersion = '1.4.1'
     jlineVersion = '2.14.5'
