From: AarjavP <AarjavP@users.noreply.github.com>
Date: Wed, 30 Aug 2017 17:39:44 +0200
Subject: [PATCH] Fix wildcard arg handling in startGroovy.bat (closes #507)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Aarjav Patel
---
 src/bin/startGroovy.bat | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/bin/startGroovy.bat b/src/bin/startGroovy.bat
index 1b7988218..f869ff7938b5a4262be297007178ef6f68aafaff 100644
--- a/src/bin/startGroovy.bat
+++ b/src/bin/startGroovy.bat
@@ -177,7 +177,7 @@
 goto process_arg
 
 :get_arg
-rem remove quotes around first arg
+rem remove quotes around first arg and don't expand wildcards
 for /f %%i in (%1) do set _ARG=%_ARG% %%~i
 rem set the remaining args
 set _ARGS=%2
